src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,224 @@
+        }
+
+        Debug.Assert(_factory.TopLevelMethod is not null);
+        Debug.Assert(_factory.CurrentFunction is not null);
+
+        // static constructors can only be invoked once, so there is no need to probe:
+        if (isMethodBody && _factory.TopLevelMethod.MethodKind == MethodKind.StaticConstructor)
+        {
+            return;
+        }
+
+        instrumentation = _factory.CombineInstrumentation(
+            instrumentation,
+            prologue: _factory.ExpressionStatement(_factory.ThrowIfModuleCancellationRequested()));
+    }
+
+    private BoundExpression InstrumentExpression(BoundExpression expression)
+        => _factory.Sequence([], [_factory.ThrowIfModuleCancellationRequested()], expression);
+
+    private BoundStatement InstrumentStatement(BoundStatement statement)
+        => _factory.StatementList(_factory.ExpressionStatement(_factory.ThrowIfModuleCancellationRequested()), statement);
+
+    public override BoundExpression InstrumentWhileStatementCondition(BoundWhileStatement original, BoundExpression rewrittenCondition, SyntheticBoundNodeFactory factory)
+        => InstrumentExpression(base.InstrumentWhileStatementCondition(original, rewrittenCondition, factory));
+
+    public override BoundExpression InstrumentDoStatementCondition(BoundDoStatement original, BoundExpression rewrittenCondition, SyntheticBoundNodeFactory factory)
+        => InstrumentExpression(base.InstrumentDoStatementCondition(original, rewrittenCondition, factory));
+
+    public override BoundExpression InstrumentForStatementCondition(BoundForStatement original, BoundExpression rewrittenCondition, SyntheticBoundNodeFactory factory)
+        => InstrumentExpression(base.InstrumentForStatementCondition(original, rewrittenCondition, factory));
+
+    public override BoundStatement InstrumentForStatementConditionalGotoStartOrBreak(BoundForStatement original, BoundStatement branchBack)
+        => InstrumentStatement(base.InstrumentForStatementConditionalGotoStartOrBreak(original, branchBack));
+
+    public override BoundStatement InstrumentForEachStatementConditionalGotoStart(BoundForEachStatement original, BoundStatement branchBack)
+        => InstrumentStatement(base.InstrumentForEachStatementConditionalGotoStart(original, branchBack));
+
+    public override BoundStatement InstrumentGotoStatement(BoundGotoStatement original, BoundStatement rewritten)
+        => InstrumentStatement(base.InstrumentGotoStatement(original, rewritten));
+
+    public override void InterceptCallAndAdjustArguments(
+        ref MethodSymbol method,
+        ref BoundExpression? receiver,
+        ref ImmutableArray<BoundExpression> arguments,
+        ref ImmutableArray<RefKind> argumentRefKindsOpt)
+    {
+        Previous.InterceptCallAndAdjustArguments(ref method, ref receiver, ref arguments, ref argumentRefKindsOpt);
+
+        // If the target method is defined within this compilation it is already being instrumented to be cancellable.
```

Comment:
>this compilation [](http://example.com/codeflow?start=50&length=16)

I think the intention was to say "this module". Methods in added modules of the same compilation are not going to be instrumented.